### PR TITLE
Fix for lambda parameter type inference in completion

### LINKED ISSUE #40216: Completion incorrectly infers lambda parameter types when the lambda is passed as a named argument in a different position than the corresponding parameter
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/825326/vs2019-intellisense-wrong.html)._

---
I have a post on stackoverflow a bout this problem. Please review it
https://stackoverflow.com/questions/58926131/graphql-netcore-visualstudio-2019-wrong-suggestions

---
#### Visual Studio Feedback System on 11/18/2019, 10:25 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 12/3/2019, 05:11 PM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes #40216

Completion was incorrectly inferring lambda parameter types when the lambda was passed as a named argument in a different position than the corresponding parameter. We now use the name of the named argument as a first lookup to find the matching parameter, followed by a positional match.

### CODE REVIEW

**Comment by dpoeschl** [comment]
Tag @jasonmalinowski as follow-up from our earlier conversation.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 1]
```diff
@@ -161,6 +147,39 @@ private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candi
             return builder.ToImmutableAndFree().Distinct();
         }
 
+        private bool TryGetMatchingParameterTypeForArgument(IMethodSymbol method, string argumentName, int ordinalInInvocation, out ITypeSymbol parameterType)
+        {
+            if (!string.IsNullOrEmpty(argumentName))
+            {
+                parameterType = method.Parameters.FirstOrDefault(p => p.Name == argumentName)?.Type;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: note: VB is case insensitive.

Comment (dpoeschl) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: So this is broken too? http://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic.Workspaces/Extensions/ArgumentSyntaxExtensions.vb,51

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: Definitely looks like it!

Comment (dpoeschl) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: I'm now getting the StringComparer from the SyntaxFacts.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 2]
```diff
@@ -151,7 +153,7 @@ private bool TryGetMatchingParameterTypeForArgument(IMethodSymbol method, string
         {
             if (!string.IsNullOrEmpty(argumentName))
             {
-                parameterType = method.Parameters.FirstOrDefault(p => p.Name == argumentName)?.Type;
+                parameterType = method.Parameters.FirstOrDefault(p => _stringComparerForLanguage.Equals(p.Name, argumentName))?.Type;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: 👍

File: src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb [hunk 3]
```diff
@@ -8161,6 +8161,28 @@ End Class
             Await VerifyItemExistsAsync(source, "Length")
         End Function
 
+        <Fact, Trait(Traits.Feature, Traits.Features.Completion)>
+        <WorkItem(40216, "https://github.com/dotnet/roslyn/issues/40216")>
+        Public Async Function CompletionForLambdaPassedAsNamedArgumentAtDifferentPositionFromCorrespondingParameterWithDifferentCasing() As Task
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb]: Test name isn't long enough.

**Comment by CyrusNajmabadi** [comment]
LGTM

**Comment by dpoeschl** [comment]
Test failures were legitimate. Pushed a fix.

**Pull request merged**